### PR TITLE
Fix tests for private reference with logical assignment

### PR DESCRIPTION
**Pull request by ptomato**
There were three things wrong with the 'and', 'or', and 'nullish' tests
that I added as part of #2940:

1. They were in the wrong folder (should be
   expressions/logical-assignment, not expressions/compound-assignment)
2. The tests for ||= and ??= on readonly accessor properties were
   incorrect. These assignments would short-circuit if the getter
   returned 1 as it previously did, so PutValue would not throw.
3. The tests for ||= and ??= on private methods were invalid, as a
   method always evaluates to true in a boolean context, and is not
   nullish, so these would always short-circuit.

I've removed the invalid private method cases, fixed the readonly
accessor cases, and added new templates to test the short-circuit
behaviour as well as the non-short-circuit behaviour.

Closes: #3413

### CODE REVIEW

**Comment by ptomato** [comment]
cc @davidot

File: src/logical-assignment-private/nullish.case [hunk 1]
```diff
@@ -0,0 +1,34 @@
+// Copyright 2022 Igalia S.L. All rights reserved.
+// This code is governed by the BSD license found in the LICENSE file.
+
+/*---
+templates:
```
Comment (ptomato) [inline src/logical-assignment-private/nullish.case:5]: This seems like it's a fairly brittle way to express what I wanted to achieve, which is "use all the templates in `default/` except for `method.template`. I didn't see a better way to do this, but let me know if there is.

**Review by davidot — COMMENTED** [review]
Thank you for fixing the test!

The fixed tests now seem correct to me.
However the new tests could be a little more thorough.

File: test/language/expressions/logical-assignment/left-hand-side-private-reference-data-property-short-circuit-and.js [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  compoundAssignment() {
+    return this.#field &&= false;
+  }
+  fieldValue() {
+    return this.#field;
+  }
+}
+
+const o = new C();
+assert.sameValue(o.compoundAssignment(), false, "The expression should evaluate to the short-circuit value");
```
[the remark this answers is not in the document]
Comment (ptomato) [reply test/language/expressions/logical-assignment/left-hand-side-private-reference-data-property-short-circuit-and.js]: I like this idea! I've done so in the new version.

**Review by davidot — COMMENTED** [review]
Look great now! (I would approve but I don't know if that messes up the PR workflow)
Whoops I found one last problem

File: test/language/expressions/logical-assignment/left-hand-side-private-reference-method-short-circuit-and.js [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+
+function doNotCall() {
+  throw new Test262Error("The right-hand side should not be evaluated");
+}
+
+class C {
+  #privateMethod() {}
+  compoundAssignment() {
+    return this.#privateMethod &&= doNotCall();
```
[the remark this answers is not in the document]
Comment (ptomato) [reply test/language/expressions/logical-assignment/left-hand-side-private-reference-method-short-circuit-and.js:48]: Oops, I forgot to skip this one. `&&=` on a private method can't ever short-circuit.

**Review by davidot — APPROVED** [review]
I can't find any problems anymore, thanks again for the fixes (and new tests)

File: src/logical-assignment-private/nullish.case (hunk 1 above)
Comment (rwaldron) [reply src/logical-assignment-private/nullish.case:5]: Did you try using a glob expression? `template: expr`?

Comment (ptomato) [reply src/logical-assignment-private/nullish.case:5]: I did look at that, but crafting a glob expression that signified "`default/*` except for one file in there" seemed worse.

**Pull request merged**